File: 1285-balance-a-binary-search-tree/1285-balance-a-binary-search-tree.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
    List<int> sortedList = [];

    void Inorder(TreeNode? root)
    {
        if(root != null) {
            Inorder(root.left);
            sortedList.add(root.val);
            Inorder(root.right);
        }
    }
    
    TreeNode? sortedListToBST(int start, int end)
    {
        if(start > end)
        {
            return null;
        }
        var mid = start + (end-start)~/2;

        TreeNode? node = new TreeNode(sortedList[mid], sortedListToBST(start, mid-1),  
                                    sortedListToBST(mid+1,end));

        return node;
    }
  TreeNode? balanceBST(TreeNode? root) {
    Inorder(root);
    print(sortedList);
    return sortedListToBST(0, sortedList.length-1);
  }
}

File: 1285-balance-a-binary-search-tree/1285-balance-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> sortedList = new ArrayList<>();

    // Inorder traversal to get elements in sorted order
    void inorder(TreeNode root) {
        if (root != null) {
            inorder(root.left);
            sortedList.add(root.val);
            inorder(root.right);
        }
    }
    
    // Function to convert sorted list to balanced BST
    TreeNode sortedListToBST(int start, int end) {
        if (start > end) {
            return null;
        }
        int mid = start + (end - start) / 2;
        TreeNode node = new TreeNode(
            sortedList.get(mid), 
            sortedListToBST(start, mid - 1),  
            sortedListToBST(mid + 1, end)
        );
        return node;
    }
    public TreeNode balanceBST(TreeNode root) {
        inorder(root);
        return sortedListToBST(0, sortedList.size() - 1);
    }
}